include/proposals/proposals_base.hpp

```diff
@@ -30,23 +34,38 @@ class Proposal {
 
   public:
 
-    Proposal(dao & _contract) : m_contract(_contract) {};
+    Proposal(dao & _contract) : m_contract(_contract), contract_name(_contract.get_self()) {};
     virtual ~Proposal(){};
 
-    virtual void create(std::map<std::string, VariantValue> & args) = 0;
+    virtual void create(std::map<std::string, VariantValue> & args);
+    virtual void update(std::map<std::string, VariantValue> & args);
+    virtual void cancel(std::map<std::string, VariantValue> & args);
+    virtual void evaluate(std::map<std::string, VariantValue> & args);
+    virtual void callback(std::map<std::string, VariantValue> & args);
+    virtual void stake(std::map<std::string, VariantValue> & args);
 
-    virtual void update(std::map<std::string, VariantValue> & args) = 0;
 
-    virtual void cancel(std::map<std::string, VariantValue> & args) = 0;
+    virtual name get_scope() = 0;
+    virtual name get_fund_type() = 0;
 
-    virtual void evaluate(std::map<std::string, VariantValue> & args) = 0;
 
-    virtual name get_scope() = 0;
+    virtual void check_can_vote(const name & status, const name & stage);
+    virtual bool check_prop_majority(std::map<std::string, VariantValue> & args);
+    virtual uint64_t min_stake(const asset & quantity, const name & fund);
```

Comment:
Would be nice to keep everything similar in the API

we usually use this func(name x, name y)

Why is this func(const name & x, const name & y) ?

I can see how const makes sense, as it leads to cleaner code, but the &... is this needed? Does it change anything?

Reply:
(not a c++ expert here ;) )

Reply:
Well, if all the functions in the contract use &, it can lead to a better performance, because now we're only using references and a reference only uses 8 bytes (usually), so in terms of memory passed among all the functions in the contract it's really small. In that way we can pass types that are heavier than 8 bytes like names or maps but we're only consuming 8 bytes.

The reference is probably useless in the action declaration like `ACTION dao::create()`, but it's definitely needed in the subsequent functions called inside `create`.

Reply:
Well unlike desktop systems, in smart contracts this kind of thing can actually make a difference, thanks for the explanation, let's leave it.